include/crypto/mlkem.h

```diff
@@ -29,199 +29,531 @@ typedef struct ossl_mlkem_ctx {
-                                   ossl_mlkem_ctx *mlkem_ctx);
+#  define DECLARE_MLKEM_ENCODE_PUB(v) \
+    void mlkem_name(v, _encode_public_key)( \
+        uint8_t *out, \
+        const mlkem_pubname(v) *public_key)
+DECLARE_MLKEM_ENCODE_PUB(512);
+DECLARE_MLKEM_ENCODE_PUB(768);
+DECLARE_MLKEM_ENCODE_PUB(1024);
+#  undef DECLARE_MLKEM_ENCODE_PUB
 
 /*
- * OSSL_MLKEM768_PRIVATE_KEY_BYTES is the length of the data produced by
- * |MLKEM768_marshal_private_key|.
+ * For v in (512, 768, 1024):
+ *
+ * ossl_mlkem<v>_encode_private_key() serializes |private_key| to |out| in the
+ * FIPS 203 internal format for ML-KEM-<v> private keys.
  */
-#  define OSSL_MLKEM768_PRIVATE_KEY_BYTES 2400
+#  define DECLARE_MLKEM_ENCODE_PRV(v) \
+    void mlkem_name(v, _encode_private_key)( \
+        uint8_t *out, \
+        const mlkem_prvname(v) *private_key)
+DECLARE_MLKEM_ENCODE_PRV(512);
+DECLARE_MLKEM_ENCODE_PRV(768);
+DECLARE_MLKEM_ENCODE_PRV(1024);
+#  undef DECLARE_MLKEM_ENCODE_PRV
 
 /*
- * ossl_mlkem768_marshal_public_key serializes |private_key| to |out| in the
- * standard format for ML-KEM-768 private keys. It returns one on success or
- * zero on allocation error. This format is verbose and should be avoided.
- * Private keys should be stored as seeds and parsed using
- * |ossl_mlkem768_private_key_from_seed|.
+ * For v in (512, 768, 1024):
+ *
+ * ossl_mlkem<v>_parse_private_key() parses a private key, in NIST's format for
+ * private keys, from |in| and writes the result to |out_private_key|. It
+ * returns one on success or zero on parse error or if there are trailing bytes
+ * in |in|. This format is verbose and should be avoided. Private keys should
+ * be stored as seeds and parsed using |ossl_mlkem<v>_private_key_from_seed|.
  */
-int ossl_mlkem768_marshal_private_key(uint8_t *out,
-                                      struct ossl_mlkem768_private_key *private_key);
+#  define DECLARE_MLKEM_PARSE_PRV(v) \
+    __owur int mlkem_name(v, _parse_private_key)( \
+        mlkem_prvname(v) *out_private_key, \
+        const uint8_t *in, \
+        ossl_mlkem_ctx *mlkem_ctx)
+DECLARE_MLKEM_PARSE_PRV(512);
+DECLARE_MLKEM_PARSE_PRV(768);
+DECLARE_MLKEM_PARSE_PRV(1024);
+#  undef DECLARE_MLKEM_PARSE_PRV
+
+/* No longer needed */
+#  undef mlkem_name
+#  undef mlkem_vecname
+#  undef mlkem_matname
+#  undef mlkem_pubname
+#  undef mlkem_prvname
+
+/* --- Variant parameters and associated methods */
+
+typedef struct {
+    mlkem_cbd_func *cbd1;
+    mlkem_cbd_func *cbd2;
+    size_t vector_bytes;
+    size_t prvkey_bytes;
+    size_t pubkey_bytes;
+    size_t ctext_bytes;
+    size_t puballoc;
+    size_t prvalloc;
+    int bits;
+    int rank;
+    int du;
+    int dv;
+    int secbits;
+} mlkem_vinfo;
+
+/* Retrive variant-specific parameters */
+extern const mlkem_vinfo *ossl_mlkem512_get_vinfo(void);
+extern const mlkem_vinfo *ossl_mlkem768_get_vinfo(void);
+extern const mlkem_vinfo *ossl_mlkem1024_get_vinfo(void);
```

Comment:
Not sure why these need the `extern` decorator

Reply:
Sure, no good reason... Fixed.